### PR TITLE
Validate doc links in PR/CI

### PR DESCRIPTION
**Pull request by asklar**
This adds link validation to our docs in CI (to prevent issues like #6312)
This change uncovered some other existing links that were invalid.

## Description
Run `unbroken` in CI
Details: https://www.npmjs.com/package/unbroken

### CODE REVIEW

**Comment by jonwis** [comment]
Can you also update https://github.com/microsoft/WindowsAppSDK/blob/main/docs/Coding-Guidelines/Languages-Markdown.md  with a "validate your doc links work by clicking on them" item?

**Review by riverar — CHANGES_REQUESTED** [review]

File: .github/workflows/docs.yml
Comment (asklar) [reply .github/workflows/docs.yml:31]: we're already using this tool in other repos. The user agent is necessary since some websites will reject requests that don't come from seemingly good UAs. The runner takes a few seconds to run, so I am not concerned.

**Comment by asklar** [comment]
> Can you also update https://github.com/microsoft/WindowsAppSDK/blob/main/docs/Coding-Guidelines/Languages-Markdown.md with a "validate your doc links work by clicking on them" item?

@jonwis it doesn't hurt to add it, but note that now broken links will not allow merging and will show a PR CI error, like below, so I'm not sure it's strictly necessary:

image

**Comment by riverar** [comment]
> we're already using this tool in other repos. The user agent is necessary since some websites will reject requests that don't come from seemingly good UAs. The runner takes a few seconds to run, so I am not concerned.

[It appears to only be used in three repositories](https://user-images.githubusercontent.com/475132/143786468-3546d7d6-c7d1-4427-87e6-dd84f4ec77c6.png), two of which are low-activity sample repositories. The highest activity repo is configured to continue on error. Am I missing any?

To repeat my suggestions ([as it's buried in the review comments](https://github.com/microsoft/WindowsAppSDK/pull/1851#discussion_r757949606)):
* configure a unique user-agent so these requests can be filtered out of logs, telemetry
* use a campaign ID to further assist with filtering (e.g. `WT.mc_id=asklar@microsoft.com`)

**Comment by DrusTheAxe** [comment]
> now broken links will not allow merging and will show a PR CI error

Is that a good thing?

@BenJKuhn another case to think about in the checkin/no-override-button conversation

**Comment by DrusTheAxe** [comment]
This is implemented in Javascript? Is there a Powershell version?

**Comment by asklar** [comment]
> > now broken links will not allow merging and will show a PR CI error
> 
> Is that a good thing?
> 
> @BenJKuhn Ben Kuhn FTE another case to think about in the checkin/no-override-button conversation

yes, it is a good thing, as it prevents making mistakes, we should encourage more of this.

**Comment by asklar** [comment]
> This is implemented in Javascript? Is there a Powershell version?

not sure why we would need to write another tool in powershell, what's wrong with it being JS?

**Comment by jonwis** [comment]
> > This is implemented in Javascript? Is there a Powershell version?
> 
> not sure why we would need to write another tool in powershell, what's wrong with it being JS?

@EricJohnson327 is working on ensuring devs can run as much of the build pipeline locally as possible, improving our inner loop.  Eric's comment (hey Eric, can you make a Discussion thread about it if one doesn't exist already?) is that yaml & js make life easy for build pipes and hard for devs.

> > Can you also update https://github.com/microsoft/WindowsAppSDK/blob/main/docs/Coding-Guidelines/Languages-Markdown.md with a "validate your doc links work by clicking on them" item?
> 
> @jonwis it doesn't hurt to add it, but note that now broken links will not allow merging and will show a PR CI error, like below, [....]

See above on inner loop. We should have documentation as to why we have the rules we have.  When someone gets the build break they should be able to self-diagnose and then understand why.  The markdown formatting guidelines say "wrap at 100, consider using a formatting tool."  Please expand it to include "**DO** ensure that all links in documentation are correct; click them in the markdown preview view."

File: docs/Coding-Guidelines/Languages-Markdown.md
Comment (asklar) [inline docs/Coding-Guidelines/Languages-Markdown.md:8]:
```suggestion
**DO** ensure that all links in documentation are correct; click them in the markdown preview view.
```

**Comment by asklar** [comment]
@jonwis thanks, I've updated the guideline for markdown with that comment.

I agree with YAML being hard for devs but disagree on the JS part. This may be a localism/lack of familiarity of some teams with JS and how most of the dev world works outside but we should embrace widely-used languages and practices.
Running the JS script locally is trivial assuming you have nodejs installed, just run `npx unbroken` in the docs directory.
To install nodejs, you can use chocolatey or winget, or download it from https://nodejs.org.

**Comment by DrusTheAxe** [comment]
> I agree with YAML being hard for devs but disagree on the JS part

'Hard' has nothing to do with it. 'Different' is the issue.

We've standardized on Powershell even at cost with zero Python _despite the benefits_ because standardization. @BenJKuhn has ES documented (yet) these principles and policies?

>assuming you have nodejs installed

That introduces another tool dependency. That's not free (dev productivity, support, compliance, ...). That's why we're being very methodical in how we manage and change our dependencies.

**Comment by DrusTheAxe** [comment]
> yes, it is a good thing, as it prevents making mistakes, we should encourage more of this.

Pardon me if I'm a bit sensitive to anything that impacts code velocity. We've been burned by that multiple times. GitHub has no 'Override' button like ADO repos can. We CANNOT checkin unless a PR passes all CI criteria<sup>1</sup>. This would make any broken URL a blocker - including infrastructure and other transient issues.

Yes, I'd like this - but I'm hesitant to endorse it as a fatal hard block.

<sup>1</sup> Or you're one of the pair admins with superpowers to blow past them.

**Comment by asklar** [comment]
> > yes, it is a good thing, as it prevents making mistakes, we should encourage more of this.
> 
> Pardon me if I'm a bit sensitive to anything that impacts code velocity. We've been burned by that multiple times. GitHub has no 'Override' button like ADO repos can. We CANNOT checkin unless a PR passes all CI criteria1. This would make any broken URL a blocker - including infrastructure and other transient issues.
> 
> Yes, I'd like this - but I'm hesitant to endorse it as a fatal hard block.
> 
> 1 Or you're one of the pair admins with superpowers to blow past them.

This pipeline runs in parallel (and even before) the ADO pipeline, and completes within seconds. The ADO pipelines take a long time (30 min or so?).

I don't see why we would _choose_ to check in a URL that is broken.
If you need to override, you can always add an exclusion for the particular URL that fails to resolve, but you have to make a conscious decision that it's ok to check in a URL that doesn't work.

**Comment by BenJKuhn** [comment]
There are a few concerns in here and I'm trying to catch up. Let me address them each in turn.

First, we're moving towards a model where everything (aside from signing) can be done locally, not just in the pipelines. Up to this point I've strongly encouraged powershell (exceptions only if strongly justified by a unique library or capability). Part of the reason for this is to simplify the developer experience. Partially it's project manageability. Any dependencies we have on build tools must be kept up to date and secure. Any dependencies we have, developers need to install as well. Many developers use node, but requiring it for one optional build step is heavy-handed, and I don't think this scenario justifies the expanded dependency graph.

Second, Howard correctly notes the issue with check-in velocity. A site going offline or content moving could break check-ins indefinitely until someone fixes and everyone else merges up. That's also problematic. It could be particularly painful in servicing branches as they age.

Third, there's also the issue of scale. If we have enough people running enough CI builds, eventually we're going to start getting the bot treatment from some sites, and the checker will start failing on otherwise valid links. That could also create headaches. 

@asklar, I don't disagree with the principal of link checking and I love the enthusiasm. A more appropriate way to implement this might be as a "background" pipeline. If you want to go down this road, set up an ADO pipeline that runs against all of the contributing repos (they can all be resources of the same pipeline, Maestro-aware to check the right branches together, public & private, etc.) and it can run once a day, filing issues / bugs as appropriate. I'm happy to talk more about what that looks like.

This ensures that no one needs to run it locally, an unrelated web-site outage never blocks code-flow, and we're not going to spamming links from an Azure server farm potentially hundreds of times a day. It still introduces a node dependency, which I'm not fond of, but if we can scope it to only being required by an infrastructure task as opposed to every developer, I think that's tolerable. Even then, I'd still prefer a powershell version (or anything else that works within existing dependencies) if there's such an option available. I wouldn't re-build this checker if it's the best thing available though.

Ben

**Comment by BenJKuhn** [comment]
One other reasonable way to implement this would be a as comment bot. I haven't built one for github, but I know they exist. Instead of failing the check it would open comments on the PR indicating which links failed, and providing instructions to edit an "ignore file" if you really want the checker to skip that link forever. Then the author (and reviewers) can decide whether to push through, apply a fix, or add an exclusion. The PR is never directly blocked on success of the link resolution though.

**Comment by riverar** [comment]
> This adds link validation to our docs in CI (to prevent issues like #6312)
> This change uncovered some other existing links that were invalid.

This issue doesn't exist, typo?

**Pull request closed without merging**
